### ISSUE TITLE
Avoid the use of ENV VARs

### ISSUE THREAD

**Issue by snim2** [softdevteam/eco#40]
The "export" functions for polyglot languages often depend on environment variables being set. It would be nicer for the user if these could be set in a separate tab of the settings dialog.

**Comment by ltratt** [comment]
Yes, I think we're crying out for this! The only thing that's not completely easy is that saving "project" settings (e.g. filenames) in the .eco file itself will mean that it's not easy to pass it from machine to machine. So perhaps we can do something like save settings per hostname? i.e. perhaps the file can have multiple project settings in it, but we can separate them by user machine so that my filenames don't squash yours and vice versa.

**Comment by snim2** [comment]
At the moment we save settings per user, I think that's probably enough to be honest. Most of the environment variables are things like directories to the root of an interpreter build, which I would guess will change between users, but won't change between projects.

**Comment by ltratt** [comment]
I might be an idiot here -- at the very least, I'm ignorant of how this bit of code works! -- but how do we save things per user?

What I was vaguely thinking of is that inside an .eco file, we could have a `settings` dict which has as its key a user hostname and its value the actual settings. So we might have something like:

```
{"hostname1" : {"UNIPYCATION" : "/usr/blah"},
 "hostname2" : {"UNIPYCATION" : "/usr/local/blah"}
}
```

where `hostname1` is your machine and `hostname2` is mine.

**Comment by ptersilie** [comment]
We are using Qt to store settings (e.g. Fonts, Theme, etc). We should use the same to store ENV VARS. No need to have different hostnames for that, tbh.

**Comment by snim2** [comment]
The piece of information you are missing is how PyQt manages settings. We have a settings dialog box here:

https://github.com/softdevteam/eco/blob/master/lib/eco/eco.py#L281-L395

and any part of the code can access the settings which are saved by it, simply by creating a configuration object and querying it. That way the "settings" code is not dependent on any other part of the code. This is how we set the colours in the heatmap, the fonts, and so on. Most of the work is done by the  PyQt4 libraries, so this is easy to do, it's just more work. The advantage of using Qt is that my settings don't need to know about yours, so the per-hostname part becomes irrelevant, the app just loads whatever settings it finds.

**Comment by ltratt** [comment]
My point is, environment variables suck as a way of passing this information to Eco. Half the time we forget and then have to quit it and reload. So we would rather have a config option saying "The PyHyp executable is at path _p_". But if we record _p_ in the Eco file, we're in trouble, because your _p_ is probably different from my _p_. So if we allow some/all settings to be per hostname, the same Eco file could record your _p_ and my _p_ in it.

I'm happy to think about alternative ways of doing this. Maybe some other bit of software has a cunning solution that we can pilfer?

**Comment by snim2** [comment]
Happily, the Settings menu, as we currently use it, provides exactly this
feature, without having to add anything to the .eco files.

**Comment by ltratt** [comment]
Ah, so there are Eco-wide settings _and_ per-project settings?

**Comment by ptersilie** [comment]
You are trying to over-engineer things again. ;) To clarify a few things: Eco at the moment does what pretty much every single software for Unix currently does. Which is storing stuff inside `~/.config`. At the moment these are only Eco wide settings. There are no project-only settings currently. But for now we should be good with just Eco-wide settings.

**Comment by snim2** [comment]
No, there are per-user settings on a given machine. So your settings on a
given machine should be fine for that host. When you move machines you will
pick up your settings on that host.

**Comment by ptersilie** [comment]
I see where the confusion was now. Laurie assumed that settings are stored per project (which means inside the eco files). Which is why he tried to come up with a solution for people having different paths. I don't think we should store the path to the executables inside the eco files. We either store them eco-wide, or we do something similar to eclipse, and generate a `.project` file for each project containing these infos, which is not supposed to be shared between hosts.

**Comment by ltratt** [comment]
Yes, I just realised that you guys were referring to the global settings (my comment died because my phone signal is poor!). I think storing the executable paths globally is a good idea. We can then store command-line options to the executables in each .eco file -- in general, they'll be pretty cross-platform.

**Comment by ltratt** [comment]
[The reason for the per-file things is that some of the Graal stuff seems to require per-file configuration, at least from what I've seen so far. I might be wrong.]

**Comment by snim2** [comment]
I haven't yet used anything per-file yet (famous last words...)

**Comment by ltratt** [comment]
:)

**Comment by ptersilie** [comment]
Fixed in #65

**Comment by ltratt** [comment]
And merged!
